snap/local/runtime-helpers/bin/setup-nginx.sh

```diff
@@ -14,3 +14,241 @@ if test ! -f "${keyfile}" ; then
+: ${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWEDMETHODS:=`yq -r .all-services.Service.CORSConfiguration.CORSAllowedMethods $SNAP_DATA/config/core-common-config-bootstrapper/res/configuration.yaml`}
+: ${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWEDHEADERS:=`yq -r .all-services.Service.CORSConfiguration.CORSAllowedHeaders $SNAP_DATA/config/core-common-config-bootstrapper/res/configuration.yaml`}
+: ${EDGEX_SERVICE_CORSCONFIGURATION_CORSEXPOSEHEADERS:=`yq -r .all-services.Service.CORSConfiguration.CORSExposeHeaders $SNAP_DATA/config/core-common-config-bootstrapper/res/configuration.yaml`}
+: ${EDGEX_SERVICE_CORSCONFIGURATION_CORSMAXAGE:=`yq -r .all-services.Service.CORSConfiguration.CORSMaxAge $SNAP_DATA/config/core-common-config-bootstrapper/res/configuration.yaml`}
+
+echo "$(date) CORS settings dump ..."
+( set | grep EDGEX_SERVICE_CORSCONFIGURATION ) || true
+
+corssnippet=/tmp/cors.block.$$
+touch "${corssnippet}"
+if test "${EDGEX_SERVICE_CORSCONFIGURATION_ENABLECORS}" = "true"; then
+  echo "      if (\$request_method = 'OPTIONS') {" >> "${corssnippet}"
+  echo "        add_header 'Access-Control-Allow-Origin' '${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWEDORIGIN}';" >> "${corssnippet}"
+  echo "        add_header 'Access-Control-Allow-Methods' '${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWEDMETHODS}';" >> "${corssnippet}"
+  echo "        add_header 'Access-Control-Allow-Headers' '${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWEDHEADERS}';" >> "${corssnippet}"
+  # Access-Control-Expose-Headers should not be set on OPTIONS request
+  if test "${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWCREDENTIALS}" = "true"; then
+    # CORS specificaiton says that if not true, omit the header entirely
+    echo "        add_header 'Access-Control-Allow-Credentials' '${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWCREDENTIALS}';" >> "${corssnippet}"
+  fi
+  echo "        add_header 'Access-Control-Max-Age' ${EDGEX_SERVICE_CORSCONFIGURATION_CORSMAXAGE};" >> "${corssnippet}"
+  echo "        add_header 'Vary' 'origin';" >> "${corssnippet}"
+  echo "        add_header 'Content-Type' 'text/plain; charset=utf-8';" >> "${corssnippet}"
+  echo "        add_header 'Content-Length' 0;" >> "${corssnippet}"
+  echo "        return 204;" >> "${corssnippet}"
+  echo "      }" >> "${corssnippet}"
+  echo "      if (\$request_method != 'OPTIONS') {" >> "${corssnippet}"
+  # Always add headers regardless of response code.  Omit preflight-related headers (allow-methods, allow-headers, allow-credentials, max-age)
+  echo "        add_header 'Access-Control-Allow-Origin' '${EDGEX_SERVICE_CORSCONFIGURATION_CORSALLOWEDORIGIN}' always;" >> "${corssnippet}"
+  echo "        add_header 'Access-Control-Expose-Headers' '${EDGEX_SERVICE_CORSCONFIGURATION_CORSEXPOSEHEADERS}' always;" >> "${corssnippet}"
+  echo "        add_header 'Vary' 'origin' always;" >> "${corssnippet}"
+  echo "      }" >> "${corssnippet}"
+  echo "" >> "${corssnippet}"
+fi
```

Comment:
Trying to understand why we extract from common config and set in the Nginx. Don't the services already set these headers? If yes, Nginx should just pass those headers to the caller.

Reply:
Thanks for explaining.

3. Can't we pass through all preflight CORS requests to services?
4. Why not skip this step and forward the backend headers to the caller along with the body as in step 8?

Reply:
> 3. Can't we pass through all preflight CORS requests to services?
> 4. Why not skip this step and forward the backend headers to the caller along with the body as in step 8?

A combination of things:

a. To pass the CORS request to the service we would have to disable authentication at the proxy layer, which removes the catch-all case.  This would leave exposed any endpoints at the microservice level that didn't implement authentication.  For example, /api/v3/ping, any custom routes the user implemented without authentication, and the rules engine REST API if that is exposed.

b. The browser will withhold sending authentication information until it has performed the preflight request and confirmed that any authorization headers are allowed to be sent.

The combination of the above means we have to handle CORS as the proxy layer.

Reply:
Not sure that I fully understand the problem. What I meant was to pass through OPTIONS requests to the services which already do respond correctly to them. This won't expose anything sensitive. 

IMO, the proxy could check the authentication header for requests others than OPTIONS and not deal with CORS at all if the backend services already do it correctly.